certora/specs/Tokens.spec

```diff
@@ -0,0 +1,137 @@
+
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+
+    function _.supply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) external => summarySupply(marketParams, assets, shares, onBehalf, data) expect (uint256, uint256) ALL;
+    function _.withdraw(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) external => summaryWithdraw(marketParams, assets, shares, onBehalf, receiver) expect (uint256, uint256) ALL;
+    function _.idToMarketParams(MetaMorphoHarness.Id id) external => summaryIdToMarketParams(id) expect MetaMorphoHarness.MarketParams ALL;
+
+    function _.expectedSupplyAssets(MetaMorphoHarness.MarketParams, address) external => NONDET;
+    function _.borrowRate(MetaMorphoHarness.MarketParams, MetaMorphoHarness.Market) external => NONDET;
+}
+
+function summaryIdToMarketParams(MetaMorphoHarness.Id id) returns MetaMorphoHarness.MarketParams {
+    MetaMorphoHarness.MarketParams marketParams;
+    uint256 lastUpdated = Morpho.lastUpdate(id);
+
+    // Safe require because markets in the supply/withdraw queue have positive last update (see LastUpdated.spec).
+    require lastUpdated > 0;
+    // Safe require because it is a verified invariant in Morpho Blue.
+    require lastUpdated > 0 => Morpho.libId(marketParams) == id;
+
+    return marketParams;
+}
+
+function summarySupply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) returns (uint256, uint256) {
+    assert shares == 0;
+    assert onBehalf == currentContract;
+    assert data.length == 0;
+
+    // Safe require because it is a verified invariant.
+    require hasSupplyCapIsEnabled(Morpho.libId(marketParams));
+    // Safe require because it is a verified invariant.
+    require isEnabledHasConsistentAsset(marketParams);
+
+    // Summarize supply as just a transfer for the purpose of this specification file, which is sound because only the properties about tokens are verified in this file.
+    Util.safeTransferFrom(marketParams.loanToken, currentContract, MORPHO(), assets);
+
+    return (assets, shares);
+}
+
+function summaryWithdraw(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) returns (uint256, uint256) {
+    assert onBehalf == currentContract;
+    assert receiver == currentContract;
+
+    MetaMorphoHarness.Id id = Morpho.libId(marketParams);
+    uint256 withdrawn = Util.withdrawnAssets(MORPHO(), id, assets, shares);
+
+    bool enabled;
+    _, enabled, _ = config(id);
+    // Safe require because:
+    // - for reallocate this is checked in the code
+    // - for withdraw, it is verified that markets in the withdraw queue are enabled.
+    require enabled;
```

Comment:
I don't understand logically why do we need to require this in order to prove what we want to prove. Even if Metamorpho was supplying or withdrawing from non enabled markets, the token accounting would still be ok no?

Reply:
We need it in conjunction with the next invariant. When we have both, we know that the market has a consistent asset, which is useful to show that we have proper accounting. Otherwise we could have a transfer of a token A from the user to MM and then a transfer of token B from MM to Morpho

Reply:
wow indeed, nice